samples/sample_common/src/d3d_allocator.cpp

```diff
@@ -20,3 +20,415 @@ or https://software.intel.com/en-us/media-client-solutions-support.
+{
+    HRESULT hr;
+
+    MSDK_CHECK_POINTER(request, MFX_ERR_NULL_PTR);
+    if (request->NumFrameSuggested == 0)
+        return MFX_ERR_UNKNOWN;
+
+    D3DFORMAT format = ConvertMfxFourccToD3dFormat(request->Info.FourCC);
+
+    if (format == D3DFMT_UNKNOWN)
+    {
+        msdk_printf(MSDK_STRING("D3D Allocator: invalid fourcc is provided (%#X), exitting\n"),request->Info.FourCC);
+        return MFX_ERR_UNSUPPORTED;
+    }
+
+    DWORD   target;
+
+    if (MFX_MEMTYPE_DXVA2_DECODER_TARGET & request->Type)
+    {
+        target = DXVA2_VideoDecoderRenderTarget;
+    }
+    else if (MFX_MEMTYPE_DXVA2_PROCESSOR_TARGET & request->Type)
+    {
+        target = DXVA2_VideoProcessorRenderTarget;
+    }
+    else
+        return MFX_ERR_UNSUPPORTED;
+
+    IDirectXVideoAccelerationService* videoService = NULL;
+
+    if (target == DXVA2_VideoProcessorRenderTarget) {
+        if (!m_hProcessor) {
+            hr = m_manager->OpenDeviceHandle(&m_hProcessor);
+            if (FAILED(hr))
+                return MFX_ERR_MEMORY_ALLOC;
+
+            hr = m_manager->GetVideoService(m_hProcessor, IID_IDirectXVideoProcessorService, (void**)&m_processorService);
+            if (FAILED(hr))
+                return MFX_ERR_MEMORY_ALLOC;
+        }
+        videoService = m_processorService;
+    }
+    else {
+        if (!m_hDecoder)
+        {
+            hr = m_manager->OpenDeviceHandle(&m_hDecoder);
+            if (FAILED(hr))
+                return MFX_ERR_MEMORY_ALLOC;
+
+            hr = m_manager->GetVideoService(m_hDecoder, IID_IDirectXVideoDecoderService, (void**)&m_decoderService);
+            if (FAILED(hr))
+                return MFX_ERR_MEMORY_ALLOC;
+        }
+        videoService = m_decoderService;
+    }
+
+    mfxHDLPair **dxMidPtrs = (mfxHDLPair**)calloc(request->NumFrameSuggested, sizeof(mfxHDLPair*));
+    if (!dxMidPtrs)
+        return MFX_ERR_MEMORY_ALLOC;
+
+    for (int i = 0; i < request->NumFrameSuggested; i++)
+    {
+        dxMidPtrs[i] = (mfxHDLPair*)calloc(1, sizeof(mfxHDLPair));
+        if (!dxMidPtrs[i])
+        {
+            DeallocateMids(dxMidPtrs, i);
+            return MFX_ERR_MEMORY_ALLOC;
+        }
+    }
+
+    response->mids = (mfxMemId*)dxMidPtrs;
+    response->NumFrameActual = request->NumFrameSuggested;
+
+    if (request->Type & MFX_MEMTYPE_EXTERNAL_FRAME) {
+        for (int i = 0; i < request->NumFrameSuggested; i++) {
+            hr = videoService->CreateSurface(request->Info.Width, request->Info.Height, 0,  format,
+                D3DPOOL_DEFAULT, m_surfaceUsage, target, (IDirect3DSurface9**)&dxMidPtrs[i]->first, &dxMidPtrs[i]->second);
+            if (FAILED(hr)) {
+                ReleaseResponse(response);
+                return MFX_ERR_MEMORY_ALLOC;
+            }
+        }
+    } else {
+        safe_array<IDirect3DSurface9*> dxSrf(new IDirect3DSurface9*[request->NumFrameSuggested]);
```

Comment:
Same issues as above. Windows build fails for me because safe_array is no longer defined (was removed by commit 4449756).

Reply:
Fixed